sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/CloudPubsubTransforms.java

```diff
@@ -0,0 +1,95 @@
+
+import static com.google.cloud.pubsublite.cloudpubsub.MessageTransforms.fromCpsPublishTransformer;
+import static com.google.cloud.pubsublite.cloudpubsub.MessageTransforms.toCpsPublishTransformer;
+import static com.google.cloud.pubsublite.cloudpubsub.MessageTransforms.toCpsSubscribeTransformer;
+
+import com.google.cloud.pubsublite.Message;
+import com.google.cloud.pubsublite.SequencedMessage;
+import com.google.cloud.pubsublite.cloudpubsub.KeyExtractor;
+import com.google.pubsub.v1.PubsubMessage;
+import io.grpc.StatusException;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+
+// A class providing transforms between Cloud Pub/Sub and Pub/Sub Lite message types.
+public final class CloudPubsubTransforms {
+  private CloudPubsubTransforms() {}
+
+  // Transform a collection of SequencedMessages to Cloud Pub/Sub received PubsubMessages.
+  public static PTransform<PCollection<? extends SequencedMessage>, PCollection<PubsubMessage>>
+      toCpsSubscribeTransform() {
+    return ParDo.of(
+        new DoFn<SequencedMessage, PubsubMessage>() {
+          @ProcessElement
+          public void processElement(
+              @Element SequencedMessage sequencedMessage, OutputReceiver<PubsubMessage> output)
+              throws StatusException {
+            output.output(toCpsSubscribeTransformer().transform(sequencedMessage));
+          }
+        });
+  }
+
+  // Transform a collection of Cloud Pub/Sub publishable PubsubMessages (ignoring message_id and
+  // publish_time) to Pub/Sub Lite Messages.
+  public static PTransform<PCollection<? extends PubsubMessage>, PCollection<Message>>
+      fromCpsPublishTransform() {
+    return ParDo.of(
+        new DoFn<PubsubMessage, Message>() {
+          @ProcessElement
+          public void processElement(@Element PubsubMessage message, OutputReceiver<Message> output)
+              throws StatusException {
+            output.output(fromCpsPublishTransformer(KeyExtractor.DEFAULT).transform(message));
+          }
+        });
+  }
+
+  // Transform a collection of Pub/Sub Lite Messages to publishab Cloud Pub/Sub incomplete,
+  // publishable
+  // PubsubMessages.
+  public static PTransform<PCollection<? extends Message>, PCollection<PubsubMessage>>
+      toCpsPublishTransform() {
+    return ParDo.of(
+        new DoFn<Message, PubsubMessage>() {
+          @ProcessElement
+          public void processElement(@Element Message message, OutputReceiver<PubsubMessage> output)
+              throws StatusException {
+            output.output(toCpsPublishTransformer().transform(message));
+          }
+        });
+  }
+
+  // Ensure that all messages that pass through can be converted to Cloud Pub/Sub messages using the
+  // standard transformation methods.
+  public static PTransform<PCollection<? extends Message>, PCollection<Message>>
```

Comment:
This will cast away the type of the input PCollection, and always return PCollection<Message>, which seems like a bad consequence of something that is intended to be a passthrough.  Maybe give this a type param?

Reply:
I don't know of a way to do this? Can you describe for me what you'd want the function signature to be?